SMPyBandits/Policies/RAWUCB.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import division, print_function  # Python 2 compatibility
+
+__author__ = "Julien Seznec"
+__version__ = "0.1"
+
+from math import sqrt, log
+import numpy as np
+import pandas as pd
+
+np.seterr(divide='ignore')  # XXX dangerous in general, controlled here!
+
+try:
+    from .BasePolicy import BasePolicy
+    from .IndexPolicy import IndexPolicy
+    from .FEWA import FEWA, EFF_FEWA
+    from .kullback import klucbBern
+except ImportError:
+    from BasePolicy import BasePolicy
+    from FEWA import FEWA, EFF_FEWA
+    from kullback import klucbBern
+    from IndexPolicy import IndexPolicy
+
+
+class EFF_RAWUCB(EFF_FEWA):
+    """
+    Efficient mechanism as described in [Seznec et al.,  2019a] (m=2) and [Seznec et al.,  2019b] (m<=2)
+    Algorithm Rotting Adaptive Window Upper Confidence Bound (RAW-UCB) [Seznec et al.,  2019b]
+    We use the confidence level \delta_t = \frac{1}{t^\alpha}.
+    """
+
+    def choice(self):
+        not_selected = np.where(self.pulls == 0)[0]
+        if len(not_selected):
+            return not_selected[0]
+        self.ucb = self._compute_ucb()
+        return np.nanmin(self.ucb, axis=1).argmax()
+
+    def _compute_ucb(self):
+        delta_t_inv = self._confidence_level_inv()
+        return (self.statistics[0, :, :] / self.windows + self.outlogconst * np.sqrt(np.log(delta_t_inv)))
+
+    def _append_thresholds(self, w):
+        # FEWA use two confidence bounds. Hence, the outlogconst is twice smaller for RAWUCB
+        return sqrt(2 * self.subgaussian ** 2 / w)
+
+    def __str__(self):
+        return r"EFF_RAW-UCB($\alpha={:.3g}, \, m={:.3g}$)".format(self.alpha, self.grid)
```

Comment:
Same, remove the `\,` as it's useless?

Reply:
I don't think so. It displays a space in my figure.